include/net/socket.h

```diff
@@ -145,6 +145,39 @@ struct zsock_addrinfo {
 	char _ai_canonname[DNS_MAX_NAME_SIZE + 1];
 };
 
+/**
+ * @brief Obtain a file descriptor's associated net context
+ *
+ * With CONFIG_USERSPACE enabled, the kernel's object permission system
+ * applies to the net_context kernel objects that underlie socket file
+ * descriptors. When a socket is opened, by default only the caller has
+ * permission, access by other threads will fail.
+ *
+ * This API is intended for threads that need to grant access to a net_context
+ * associated with a particular file descriptor to another thread. The
+ * returned  net_context pointer represents the underlying kernel object and
+ * may be passed to APIs like k_object_access_grant().
+ *
+ * In a system like Linux which has the notion of threads running in processes
+ * in a shared virtual address space, this sort of management is unnecessary as
+ * the scope of file descriptors is implemented at the process level.
+ *
+ * However in Zephyr the file descriptor scope is global, and MPU-based systems
+ * are not able to implement a process-like model due to the lack of memory
+ * virtualization hardware. They use discrete object permissions and memory
+ * domains instead to define thread access scope.
+ *
+ * User threads will have no direct access to the returned net_context
+ * and will fault if they try to access its memory; the pointer can only be
+ * used to make permission assignment calls, which follow exactly the rules
+ * for other kernel objects like device drivers and IPC.
+ *
+ * @param sock file descriptor
+ * @return pointer to associated struct net_context, or NULL if the
+ *         file descriptor wasn't valid or the caller had no permission
+ */
+__syscall struct net_context *zsock_get_context_object(int sock);
```

Comment:
this is wrong, socket objects are `void *`, I got confused from older code where everything was backed by a `net_context` but it's much more abstracted now. Will fix this and other changes that coerced fd object types into a `net_context` pointer, needs to be left as `void *`.